src/pybind/mgr/dashboard/services/hardware.py

```diff
@@ -0,0 +1,91 @@
+
+
+import json
+from typing import Any, Dict, List, Optional
+
+from ..exceptions import DashboardException
+from ..services.orchestrator import OrchClient
+
+
+class HardwareService(object):
+
+    @staticmethod
+    def get_health(categories: Optional[List[str]] = None,
+                   hostname: Optional[List[str]] = None):
+        total_count = {'total': 0, 'ok': 0, 'error': 0}
+
+        output: Dict[str, Any] = {
+            'total': {
+                'category': {},
+                'total': {}
+            },
+            'host': {
+                'flawed': 0
+            }
+        }
+
+        categories = HardwareService.validate_categories(categories)
+
+        orch_hardware_instance = OrchClient.instance().hardware
+        for category in categories:
+            data = json.loads(orch_hardware_instance.common(category, hostname))
+            category_total = {'total': 0, 'ok': 0, 'error': 0}
+
+            for host, items in data.items():
+                count_items = {'total': 0, 'ok': 0, 'error': 0}
+
+                if host not in output['host']:
+                    output['host'][host] = {'flawed': False}
+                for item in items.values():
+                    if item['status']['health'] == 'OK':
+                        count_items['ok'] += 1
+                    else:
+                        output['host'][host]['flawed'] = True
+
+                    count_items['total'] += 1
+                count_items['error'] = count_items['total'] - count_items['ok']
+                # Update items by host's category
+                output['host'][host][category] = count_items
+
+                # Update category total items
+                category_total['total'] += count_items['total']
+                category_total['ok'] += count_items['ok']
+                category_total['error'] += count_items['error']
+
+            if category not in output['total']['category']:
+                output['total']['category'][category] = {}
+            output['total']['category'][category] = category_total
+
+            # Update overall total counts
+            total_count['total'] += category_total['total']
+            total_count['ok'] += category_total['ok']
+            total_count['error'] += category_total['error']
+
+        for host in output['host']:
+            if host == 'flawed':
+                continue
+            if output['host'][host]['flawed']:
+                output['host']['flawed'] += 1
+
+        output['total']['total'] = total_count
+
+        return output
```

Comment:
i tried to simplify this to something like this
```
category_total = {
            'total': sum(len(items) for items in data.values()),
            'ok': sum(item['status']['health'] == 'OK' for items in data.values() for item in items.values()),
            'error': 0
        }

        for host, items in data.items():
            output['host'].setdefault(host, {'flawed': False})
            output['host'][host]['flawed'] = any(item['status']['health'] != 'OK' for item in items.values())

        category_total['error'] = category_total['total'] - category_total['ok']
        output['total']['category'].setdefault(category, {})
        output['total']['category'][category] = category_total

        total_count['total'] += category_total['total']
        total_count['ok'] += category_total['ok']
        total_count['error'] += category_total['error']

    output['total']['total'] = total_count

    output['host']['flawed'] = sum(1 for host in output['host'] if host != 'flawed' and output['host'][host]['flawed'])

    return output
```

this is also more or less what you did but just taking advantage of the list comprehension.

Reply:
it might need testing though, but if its too much to do we can ignore it for now

Reply:
Sure, I can make the changes.
I'll use `.setdefault()` to avoid unnecessary comprobations and make use of the comprehensions lists to make the code more concise. 

For future references, I'll try to avoid nesting for loops and use comprehensions list when possible.